magento/magento2
# Cron cleanup repeatedly hits deadlocks on large environments where groups can overlap

We have two large installations, one Enterprise, one Community, both on 2.3.4, with some long running crons. Frequently the groups overlap and the DELETE operations in the cron cleanup cause deadlocks.

Examining the database environment we can see that some DELETE queries are using the job_code index whereas other are performing a scan as they attempt to use the status column. The ones using an index take exclusive locks on the index then proceed to filter, release, and lock the relevant PRIMARY key entries to take the rows. The queries not using the index take the PRIMARY key locks first then perform filtering and release them. This may not be an exact description but it's my current understanding of what's happening. As you can logically see when two queries run at the same time for two different groups it can cause a deadlock.

### Description (*)
I've tested this patch over time and it alleviates the deadlocks I'm seeing.

The patch changes the DELETE for cleanup so it does not use a range for the job_code. This makes the query more likely to use the job_code index, mostly on MySQL 5.7. However, depending on database contents, it can still ignore it if it finds status column is better, mostly on MySQL 5.6 which I tested too. Adding the status column to the job_code index resolves this and I have yet to see incorrect index behaviour.

The additional index also helps with the atomic job status update. Previously the left join to get the existing job status would not always use an index. And could deadlock also with the aforementioned DELETE cleanup queries. The new index on job_code and status makes this join always use the index.

Furthermore, I had issues on development environments which were built from snapshots, and frequently shutdown and restarted, where there would accumulate "running" jobs that never finish as those jobs were terminated whilst running, or were running at the time of the snapshot. This PR also adds a cleanup of those so that after 24 hours they become failed. This matches the logic in the atomic status update, which ignores running jobs older than 24 hours, allowing a new job to start (and was the cause of the buildup, as every day there was a chance of one additional running entry per job.)

### Related Pull Requests
1. This PR provides alternative fix for cleaning up running jobs that failed to the fix in this one: magento/magento2#24789: magento/magento2#23054:
2. This PR adds an index similar to PR. Not sure what issue was in this PR but it could be the UPDATE/LEFT JOIN slowness is similar as that can avoid using index without this PR magento/magento2#27391: magento/magento2#: Add a new index for `cron_schedule` table

### Fixed Issues (if relevant)
1. Does not fix OP issue but fixes issue in some comments: Fixes magento/magento2#8933 : 1213 Deadlock found when trying to get lock
2. Fixes magento/magento2#22438 : Magento 2.3.1 Cron Deadlocks for cron_schedule
3. Fixes magento/magento2#18409 : Magento 2.2.5 - Cron Job Error. SQLSTATE[40001]: Serialization failure: 1213 Deadlock
4. Fixes #23054
5. Related #25634 (High CPU can occur due to growing cron_schedule, which this PR resolves)
6. Related #26507 (High CPU can occur due to growing cron_schedule, which this PR resolves)
7. Related #26809 (High CPU can occur due to growing cron_schedule, which this PR resolves)

### Manual testing scenarios (*)
1. Unable to reproduce on demand but this steps hopefully are what we understand is the cause
2. Setup new Enterprise site and install several new modules with their own cron groups performing tasks taking at least 5 minutes and starting every 10 minutes - cron groups here are sync tasks with remote systems and stock update and product updates
3. Populate 1000s of products and 1000s of customers
3. Run cron every minute and set cron to email on failure
4. Leave running for a week, and notice emails with deadlocks

For running stale jobs:
1. Take database snapshot whilst jobs running and start it up on new environment
2. Periodically shutdown server whilst jobs are running, several times
3. Note even when cron not running, there are running jobs in the cron_schedule table
4. Note after a week these are still there

### Questions or comments
N/A

### Contribution checklist (*)
